### ISSUE TITLE
WHITESPACE_ONLY mode does non-whitespace-only optimizations

### ISSUE THREAD

**Issue by AshleyScirra** [google/closure-compiler#3637]
Closure compiler v20200614

in.js:
```js
{
	function foo(p) { self["external1"](p) }
	
	console.log(foo());
	console.log(foo());
}
```

Command: `java -jar ./closure-compiler.jar --js in.js --js_output_file out.js --compilation_level WHITESPACE_ONLY --formatting PRETTY_PRINT`

out.js:
```js
 {
  var foo = function(p) {
    self["external1"](p);
  };
  console.log(foo());
  console.log(foo());
}
;
```

Note that `function foo(p)` was rewritten as `var foo = function(p)`. This is not a whitespace-only change. It also causes a bug (#3623). Therefore even using the lowest tier of WHITESPACE_ONLY optimizations, we cannot escape this bug. I think it is reasonable to expect WHITESPACE_ONLY to only do what it says and not make other adjustments which alter the way scripts work.

**Comment by christiantjl** [comment]
Following for updates. This is expected if `language_in` and `language_out` are different (transpilation will still be necessary), but unfortunately this persists even if you manually specify the same output for both (i.e. `--language_in STABLE --language_out STABLE`).

@brad4d, I managed to make progress on running the `goog.module` runtime tests, but they all currently fail due to `WHITESPACE_ONLY` still resulting in `$jscomp` name mangling due to transpile. Will try to work around. (Non-`goog.module` tests are working as expected).

![image](https://user-images.githubusercontent.com/1657236/87844906-7195bc00-c887-11ea-8806-839eb4912a35.png)

**Comment by concavelenz** [comment]
Any language_out above ECMASCRIPT_2015 should avoid this issue.  "STABLE" as language_out is ECMASCRIPT5 currently (but for no good reason at this point).

I think that I would like to take WHITESPACE_ONLY back to its original intent: simply parse and print.  That however, would mean that this would be restricted to a single "module" file input, or any number of scripts (which can be concatenated) but not both.

**Comment by christiantjl** [comment]
@concavelenz After spending a few days diving through the source I found the `NO_TRANSPILE` language_out flag, which seems to be the magic bullet for situations like this.

**Comment by AshleyScirra** [comment]
Is that an undocumented option? [The flags and options documentation](https://github.com/google/closure-compiler/wiki/Flags-and-Options) doesn't appear to reference it.

**Comment by christiantjl** [comment]
@AshleyScirra Yeah, I only found it in passing. The only public reference I can find to it is @brad4d's issue https://github.com/google/closure-compiler/issues/3256.

**Comment by AshleyScirra** [comment]
What's the difference between NO_TRANSPILE and setting both the input and output language to the same thing?

**Comment by christiantjl** [comment]
@AshleyScirra There isn't supposed to be one - same input/output language should also imply no transpilation. It doesn't end up being 100% coverage, but it does work usually:

`$ google-closure-compiler -O WHITESPACE_ONLY --formatting PRETTY_PRINT --language_in ES_NEXT --language_out ES_NEXT`
```javascript
'use strict';
{
  function foo(p) {
    self["external1"](p);
  }
  console.log(foo());
  console.log(foo());
}
;
```

Also works for `ECMASCRIPT5` and a few others I tested. I think what the issue is here is that both `language_in` and `language_out` are set to `STABLE` by default, but `STABLE` is actually *different* for input and output (ends up being `STABLE_IN` and `STABLE_OUT` internally).

It's definitely a bug, but `--NO_TRANSPILE` output mode is a good trick to have to work around it.
